Add style variations to the Site Editor

scruffian opened the pull request:
This adds style variation creation to the Site Editor:

https://user-images.githubusercontent.com/275961/181806075-2aa918a8-88aa-4652-b52c-bce0d95b8620.mp4

This adds a new JS package that allows us to integrate this tool more deeply into the Gutenberg Site Editor.

The initial step is to add the creation of style variations, but in the future we should move across other tools.

To test you need to use this Gutenberg PR: https://github.com/WordPress/gutenberg/pull/42824

scruffian committed:
Add tools to the site editor

save the variation locally

generate a zip with the variation

shanged wording, added script to toggle between forms

wording and paths

duplicate variations with the same name instead of replacing them

revert disconnected changes

scruffian committed:
Add blank theme creation to the site editor

scruffian commented:
I also added blank theme creation to the site editor:

<img width="857" alt="Screenshot 2022-08-01 at 12 11 10" src="https://user-images.githubusercontent.com/275961/182135787-4b98644b-4516-4f05-a6d8-dce11313a39e.png">

madhusudhand commented on src/index.js:
```diff
@@ -0,0 +1,111 @@
+	return (
+		<PanelBody>
+			<MenuGroup label={ __( 'Create style variation' ) }>
+				<TextControl onChange={ ( value ) => setVariationName( value ) } placeholder={ __( 'Variation name' ) }></TextControl>
```
May be a short description can be added to give the context of the feature.

`Save user changes as a new style variation of the current theme`

MaggieCabrera replied:
I wonder if there's a tooltip component or something like that in GB so we don't add a lot fo text while still conveying all the information for the action

pbking replied:
Crashing for me too.

madhusudhand reviewed (commented):
Plugin is crashing on click of the button.

<img width="333" alt="image" src="https://user-images.githubusercontent.com/1935113/182139810-9c5292e9-fe8b-4682-a5ea-0b26d1091d6b.png">

<img width="847" alt="image" src="https://user-images.githubusercontent.com/1935113/182138771-488ea488-b710-4a94-8a79-82b3d28b1914.png">

scruffian commented:
@madhusudhand @pbking sorry I forgot to mention that you also need this associated Gutenberg PR: https://github.com/WordPress/gutenberg/pull/42824

jffng commented on src/index.js:
```diff
@@ -0,0 +1,111 @@
+
+const StyleVariations = () => {
+	const [ variationName, setVariationName ] = useState( '' );
+	const [ canReset, onReset ] = wp.editSite.useGlobalStylesReset();
```
It's nice for demo purposes but I'm wondering from a UX perspective, does it make sense to reset the styles when the variation is created?

scruffian replied:
That's what we do in the PHP side of the tool.

pbking replied:
Ideally when in this perspective we could activate the variation (which would leave all of the styles as-set instead of resetting to default).

jffng commented on src/index.js:
```diff
@@ -0,0 +1,111 @@
+
+	return (
+		<PanelBody>
+			<MenuGroup label={ __( 'Create style variation' ) }>
```
There's a padding applied to this label that I think looks off, I think it should be just left aligned, if possible. Maybe that's a problem with the component though.

<img width="278" alt="Screen Shot 2022-08-09 at 2 25 04 PM" src="https://user-images.githubusercontent.com/5375500/183734301-1962615c-982d-49e9-9aef-03893b834b85.png">

scruffian replied:
Yeah I'd rather fix that in the component itself.

jffng reviewed (commented):
This worked well for me in combination with the https://github.com/WordPress/gutenberg/pull/42824.

If Gutenberg is deactivated, the editor crashes when I click on the Create Block Theme panel, so perhaps a check should be added for whether to initialize this panel.

pbking commented:
> If Gutenberg is deactivated, the editor crashes when I click on the Create Block Theme panel, so perhaps a check should be added for whether to initialize this panel.

Yeah, definitely either need a check for capabilities or wait for that change to land in CORE to release this.

scruffian committed:
Add checks to ensure we have what we need from Gutenberg

scruffian commented:
> Yeah, definitely either need a check for capabilities or wait for that change to land in CORE to release this.

Done in dfbc74f.

jffng reviewed (approved):
Would like to see this one land since it makes designing a variation in the site editor so much easier, especially if you are switching around between versions / branches a lot.

jffng commented:
Ah just as I green checked, I noticed one bug — when I try to save a variation with a space in its name, it 404s:

<img width="427" alt="Screen Shot 2022-08-11 at 12 29 58 PM" src="https://user-images.githubusercontent.com/5375500/184183408-8f70eceb-7b37-4079-b76b-1695a7d987a0.png">

scruffian committed:
update to use unstable api

Commit:
Solve conflicts and bring PR up to date

What:
- Move Style Variations to secondary navigation screen
- Remove outdated API

Some of the removed code has to be re-worked using current APIs.i

vcanales commented:
It's been a while, so I took the liberty to bring this up to date removing some code that was breaking due to changes in the App. I also refactored it to reflect how other parts of the Plugin behave within the Site Editor. If @scruffian has no objections, I'd be happy to dedicate a bit of time to this.

vcanales commented on includes/class-create-block-theme-api.php:
```diff
@@ -105,7 +126,7 @@ function rest_get_readme_data( $request ) {
 				array(
 					'status'  => 'SUCCESS',
-					'message' => __( 'Readme file data retrieved.', 'create-block-theme' ),
+					'message' => __( 'Readme file data retrieved . ', 'create-block-theme' ),
```
It seems like an auto-formatter was a bit overzealous with its job. Will be correcting these.

pbking replied:
🤣

Commit:
remove unintended space around periods

pbking commented:
I would love to see this wrap up and land.  It was the inspiration and reference for other site editor changes but that never included variations!  Would be keen to review or assist on wrapping this up.

scruffian commented:
I'm happy for anyone to take this over.

madhusudhand commented:
closing this in favour of #532
